### ISSUE TITLE
PkgServer synchronization (Pkg Server version of General is delayed relative to Git clone of General)

### ISSUE THREAD

**Issue by johnnychen94** [JuliaRegistries/General#16777]
With pkg server enabled by default since Julia 1.5, there's an issue that a release PR is merged in General while the new version is still unavailable in the storage server in a short period. Because users don't know whether the pkg server has synced the commit, this would frequently break the CI test.

One such example is: once ImageMorphology v0.2.6 is added to General, I immediately retrigger the CI in https://github.com/JuliaImages/Images.jl/pull/895 and then CI on nightly fails because it couldn't find ImageMorphology v0.2.6. In this case, the PR merged notification is a lie to developers 😂 

cc: @staticfloat @StefanKarpinski

**Comment by StefanKarpinski** [comment]
The Pkg client should fall back to fetching the package version directly from GitHub, no?

**Comment by johnnychen94** [comment]
I think the pkg client would directly download the "latest" registry from pkg server, and then it finds no v0.2.6 when doing version resolving.

**Comment by DilumAluthge** [comment]
@staticfloat How frequently does the Pkg server update its copy of the General registry?

**Comment by StefanKarpinski** [comment]
It runs in a loop continuously pulling the registry and updating things. So it’s generally mostly up-to-date but not instantaneous.

**Comment by StefanKarpinski** [comment]
Depending on a point version that you just published in CI seems like kind of a corner case? You have to wait for registration to go through as well, how is waiting for it to get into storage servers any different?

**Comment by johnnychen94** [comment]
That I received a merged notification/email from General, I retriggered the CI, and then :bangbang: "no this version is not available yet."

Yes, this is an edge case for CI only, and it could be totally fine to do the rest of the work later. It's just not making the pipeline as smooth as it was.

**Comment by DilumAluthge** [comment]
It would be nice to get some concrete measurements on this.

If the delay between "PR merged in General" to "new version is available from the Pkg server" is 5 minutes, then I think that's no big deal.

But if the delay is e.g. 30 minutes, I think that would be annoying and should be fixed.

@johnnychen94 How long was the delay for you?

> Or should we unset PkgServer in CI?

That would deprive the community of some useful telemetry statistics. I'd rather be able to keep using Pkg Server in CI.

**Comment by johnnychen94** [comment]
I checked it again after 15mins and it failed, then I went away to do other works. I'll report if I get more data, it usually takes about 8-15mins for an incremental update in my storage server in LAN. 

The current loops in [gen_static.jl](https://github.com/JuliaPackaging/PkgServer.jl/blob/dddf7087e9428ed57fb32fafd44e1a71a50c83ad/bin/gen_static.jl) iterates on all packages and all versions, and most of the time is wasted on untaring existing versions only to get `Artifacts.toml`, but we could actually only iterate on the registry diffs.

**Comment by DilumAluthge** [comment]
> It runs in a loop continuously pulling the registry and updating things. So it’s generally mostly up-to-date but not instantaneous.

> I checked it again after 15mins and it failed, then I went away to do other of my work, now it works. I'll report if I get more data.

Personally, I think that 15 minutes is too slow. It would be good to get this delay down to 5 minutes or less, in my opinion.

It seems like the only issue here is updating the registry, right? If you have an updated registry, but the tarball of the code is not in the Pkg Server, then you'll just fall back to downloading the tarball from Git. The issue here is that the registry itself is not up to date.

Could we have two loops in the Pkg Server, running in parallel side by side? One loop does just the registry. It repeatedly updates the registry. The other loop does everything else.

**Comment by DilumAluthge** [comment]
> Personally, I think that 15 minutes is too slow. It would be good to get this delay down to 5 minutes or less, in my opinion.

Just to elaborate, in the pre-PkgServer days, the delay was essentially zero. As soon as a pull request was merged, you only had to wait a few seconds before updating your registry, since the registry was a Git repo, and GitHub updates Git repos within a matter of seconds.

So in my opinion, going from a delay of less than one minute to a delay of greater than or equal to fifteen minutes is a clear regression.

**Comment by fredrikekre** [comment]
I just tested and for me it was less than 1 minute. I don't think we can do better than that, we have to let stuff propagate through the system.

**Comment by DilumAluthge** [comment]
> I just tested and for me it was less than 1 minute. I don't think we can do better than that, we have to let stuff propagate through the system.

1 minute is definitely fine. I would say anything less than or equal to 5 minutes is fine.

**Comment by johnnychen94** [comment]
> Could we have two loops in the Pkg Server, running in parallel side by side? One loop does just the registry. It repeatedly updates the registry. The other loop does everything else.

Personally, I think this can be a good idea. But just to be clear, IIUC it's deliberately designed to update `/registries` [when update finishes](https://github.com/JuliaPackaging/PkgServer.jl/blob/dddf7087e9428ed57fb32fafd44e1a71a50c83ad/bin/gen_static.jl#L343-L347) so that the storage server "strictly" follows the completeness requirement: any data declared by the server is available in the server. If we continuously update `/registries`, it actually breaks such commitment.

**Comment by DilumAluthge** [comment]
It's curious that for Fredrik the update took 1 minute, but for Johnny it took more than 15 minutes.

It might be helpful to collect more samples. Perhaps someone could write a script that routinely pulls the registry from the Pkg server, and pulls a list of recently merged PRs from General, and loops through the recently merged PRs (starting with the most recently merged PR) and goes backwards in time until it finds the most recent PR that is included in the registry provided by Pkg server. If we automate that process, we can collect a lot of samples and figure out how common it is for the Pkg server registry to be more than 5 minutes delayed.

**Comment by DilumAluthge** [comment]
> > Could we have two loops in the Pkg Server, running in parallel side by side? One loop does just the registry. It repeatedly updates the registry. The other loop does everything else.
> 
> Personally, I think this can be a good idea in practice. But just to be clear, IIUC it's deliberately designed to update `/registries` [only when the whole update succeed](https://github.com/JuliaPackaging/PkgServer.jl/blob/dddf7087e9428ed57fb32fafd44e1a71a50c83ad/bin/gen_static.jl#L343-L347) so that the storage server "strictly" follows the completeness requirement: any data declared by the server is available in the server. If we continuously update `/registries`, it slightly breaks such commitment.

Yeah I realize now my "two loops" suggestion breaks that promise. Maybe best to look for other solutions.

**Comment by DilumAluthge** [comment]
It just seems like a bad user experience that if a bug fix is registered and merged in General, now you have to wait an unknown period of time before the bug fix is accessible to you?

At the very least, it would be useful to be able to figure out how recent your registry is. Does the Pkg server expose any endpoint that would let me get e.g. the UTC timestamp corresponding to the last time when the Pkg server cloned the registry? So at least I can get a sense of how recent my Pkg server registry is.

**Comment by StefanKarpinski** [comment]
@johnnychen94, are you running a modified version of the `gen_static.jl` script? In the default configuration, it should only be downloading artifacts that appear in newly registered package versions, which should be fast, much closer to the 1 minute for an incremental update that @fredrikekre is seeing than the 15 minutes you're seeing.

**Comment by StefanKarpinski** [comment]
E.g. do you have `get_old_package_artifacts` set to true or something?

**Comment by johnnychen94** [comment]
The 15 minutes I observed is an approximate time in CI and not using my local storage server.

> do you have get_old_package_artifacts set to true or something?

Oops, looks like I didn't set this correctly. My local storage server, however, does run a refactored-version of `gen_static.jl`, i.e., [StorageServer.jl](https://github.com/johnnychen94/StorageServer.jl).

---

I'll make a script as @DilumAluthge suggested and give a further report on this next weekend.

**Comment by johnnychen94** [comment]
The following is what I've collected last 12 hours on `us-east.storage.julialang.org`:

```csv
registry_hash, available_time, wait_time(min)
8c9af6ba0ba9dc8605ed2334ca3ef58c06438bc7, 2020-07-16T16:57:37.829, 24.447133333333333
89daa2df5a8d0623dc3dc01f2456d29bd68cc255, 2020-07-16T18:25:25.7, 40.07833333333333
e405b2ca00df3fd3de42e1598d124487065abada, 2020-07-16T18:46:30.462, 23.57435
356e277b1e3392f7eea6097dbfab68057d17b95e, 2020-07-16T19:50:36.881, 35.81466666666667
d555acd84f406dbd22565988269ee83d67fdeb7f, 2020-07-16T21:15:58.31, 30.588483333333333
f89846fab0011a3abc1d40a00fb7e9764c60597c, 2020-07-16T21:38:32.576, 23.02625
0f9b1ff36d1c1f57144b7a469b4a2f467893a6be, 2020-07-16T22:20:50.83, 23.14715
2d91520d48691a98203fa5fdce5c974affe3705c, 2020-07-16T23:04:23.309, 40.788466666666665
ac1d300e4a9e02097170df19dc2b49d3c4caa5bc, 2020-07-16T23:26:00.964, 28.81605
fae1e99903414859c680f62eeb1d85ddcd568256, 2020-07-16T23:46:10.641, 23.010666666666665
fde3f8682be0da1cafe9a43c209859b20d2a4b94, 2020-07-17T00:30:23.78, 33.57965
342b73e1e663d9b9d96714f9cf3daff5cadad2c0, 2020-07-17T00:51:15.738, 33.9456
f76f5bf5167e444bbb451e74c3c5a9c8ceb90088, 2020-07-17T01:14:09.782, 41.27968333333333
faf9767c320fae08b7f304acea08beee0ada0731, 2020-07-17T01:35:35.17, 38.48616666666667
1e4af49160e6b2a24089d8b364a137d7a04d8ced, 2020-07-17T01:56:48.785, 40.82975
a4d4bf9c53ee8b479e20bb7e26e62eda7df621a7, 2020-07-17T02:25:55.667, 40.64445
d2c616f9bcb54db4af425872f4d0ab7366541268, 2020-07-17T02:46:42.314, 23.105216666666667
a6d1bf1d78fc91c9ce6115679f3448bf381a90ef, 2020-07-17T03:51:35.29, 36.53815
c0a56ff9bf3ec1c09f5bc7f1be69b681384b1fe6, 2020-07-17T04:55:50.671, 22.161183333333334
```

Wait time: the time I found a new hash from pkg server - the time record in git commit

script: https://gist.github.com/johnnychen94/98fde55fc341d0c967f8f5ef2a48956a

**Comment by staticfloat** [comment]
That's great Johnny, we should track this over time somehow.

**Comment by staticfloat** [comment]
I've just fixed some issues with the Korean storage server; please keep track of the latency of Registry -> https://kr.storage.julialang.org over the next couple of days.  I think it shuold be much better than in the past.

One remaining design reason why the registry updates may be slow sometimes is that the storage server does not advertise the new registry hash until it has downloaded and stored all new resources; e.g. it doesn't advertise a registry until it can serve everything referenced by that registry.  It's possible we may want to change that in order to expedite registry service, but I'm not 100% sure.

**Comment by johnnychen94** [comment]
Tracking `kr.storage.julialang.org` looks great now!

```
registry_hash, available_time, wait_time(min)
a0e8965316b0cabd279a427796174f394ec8fe51, 2020-07-19T01:57:07.609, 59.66011666666667
69bde3f1b510c45ad3f6cc78c927f59f9cd0a312, 2020-07-19T02:49:48.119, 76.96865
784c3cdc16a64f605f87a63903c387647d6dbebc, 2020-07-19T15:44:43.319, 59.8053
c1b1462946f5e3eb365b498bfab10a613c8ea02c, 2020-07-19T16:34:40.143, 1.5023666666666666
6b2351fad68d1461ae86280226d92f3c2f071d2d, 2020-07-19T17:57:50.054, 0.6508833333333334
3daa85af49a9c315ed784bc84a626c5ee04d6393, 2020-07-19T18:24:17.493, 1.3748666666666667
2226550d169a9aa18cc5797655ba4890a3f67c72, 2020-07-19T18:59:04.296, 1.88825
49592182216d517575d2651aa9d9c2df9a4d2b88, 2020-07-19T20:20:04.313, 0.5052
04a9ee5adbdb1a3d40a101ec73e3a22f41da96ae, 2020-07-19T20:57:56.493, 0.8748666666666667
4bbbc9a9f3c468fb00f95ce9055640d66d31f1b8, 2020-07-20T01:34:59.707, 1.81175
9f72ff8688a1d04fa5667fe36a00910fdab79a4a, 2020-07-20T01:46:55.071, 1.5511666666666666
8d79f17d93a439ecc7a4ae9af57c3a8ea45941a7, 2020-07-20T02:58:15.375, 1.3062333333333334
b93141ba4c6bf0c2870b9a18845297a085df8ef8, 2020-07-20T04:18:33.838, 1.0305333333333333
8c56073f6bff521db16bdb7b64afd5722cb2fbe7, 2020-07-20T06:34:38.472, 1.4077
443ade04a5d52b01e2d615d2e3186f754527fa8b, 2020-07-20T08:43:06.404, 0.89005
```

**Comment by StefanKarpinski** [comment]
The storage server code is not public and is substantially more complex than the simple static server script. The premise as outlined in the original Pkg design issue, is that different entities provide independent storage services, which are treated uniformly by the pkg servers. @staticfloat and I have talked about exposing new `/packages` and `/artifacts` endpoints that list all package and artifact resources like `/registries` does for the current registries, which would make it easy to mirror everything that a storage server knows about.

**Comment by StefanKarpinski** [comment]
Also, great to see those fast update times! That's what I had always imagined this should be like. It should continue to be like that going forward.

**Comment by DilumAluthge** [comment]
> The storage server code is not public and is substantially more complex than the simple static server script.

Would it be possible to eventually open source the storage server code?

**Comment by StefanKarpinski** [comment]
That is not something we're planning. The storage servers are built and maintained by Julia Computing, offered to as a free service to the community. A large part of their functionality is interacting with proprietary systems like GitHub, GitLab to get resources and AWS/S3 for persistence and those capabilities are also key features of JC's JuliaTeam product offering. If anyone else wants to build and maintain a storage service, they should absolutely do so—the protocol is very simple. I do think that we should have an open source script that mirrors the Julia Computing storage servers and serves them statically. That will act as a backup in case the JC storage servers go down.

I'm realizing now that since the storage servers are JC proprietary they probably should not be called {us-east,kr}.storage.julialang.org but should instead be named {us-east,kr}.storage.juliahub.com. @staticfloat, how hard would it be to change their host names?

**Comment by staticfloat** [comment]
Not hard.

**Comment by DilumAluthge** [comment]
> That is not something we're planning. The storage servers are built and maintained by Julia Computing, offered to as a free service to the community. A large part of their functionality is interacting with proprietary systems like GitHub and GitLab to get resources and AWS/S3 for persistence and those capabilities are also key features of JC's JuliaTeam product offering.

That makes sense to me!

**Comment by DilumAluthge** [comment]
> I do think that we should have an open source script that mirrors the Julia Computing storage servers and serves them statically. That will act as a backup in case the JC storage servers go down.

That would be great to have.

**Comment by johnnychen94** [comment]
A regular update for the past week:

```
[ Info: general.kr.storage.julialang.org.csv
Summary Stats:
Length:         147
Missing Count:  0
Mean:           11.665101
Minimum:        0.461833
1st Quartile:   1.125500
Median:         1.720183
3rd Quartile:   5.390308
Maximum:        681.200750
Type:           Float64
[ Info: general.us-east.storage.julialang.org.csv
Summary Stats:
Length:         182
Missing Count:  0
Mean:           35.491747
Minimum:        21.210133
1st Quartile:   26.917017
Median:         33.539417
3rd Quartile:   38.245858
Maximum:        257.773317
Type:           Float64
```

**Comment by StefanKarpinski** [comment]
Could we hook this up to the status page?

**Comment by DilumAluthge** [comment]
@staticfloat @fredrikekre Would it be possible to add these "registry update times" to the status page at `status.julialang.org`?

**Comment by staticfloat** [comment]
Unfortunately not, but if someone wants to help with this project, I would be happy to set up a JuliaLang timescaleddb or influxdb server.  We can run a script that generates these lag times, store them to the time series database, then visualize it with a public grafana instance or something.

If someone would like to help with that, contact me or post here.  If noone speaks up I'll eventually get to it, but it might take me a few months.  :P

**Comment by visr** [comment]
Not sure if it is related, but ArchGDAL v0.5.0 was merged in the general registry 3 days ago ([registry PR](https://github.com/JuliaRegistries/General/pull/19593)), but I can still not install it in a clean environment, on both julia 1.4 and 1.5, and either of these two in `startup.jl`:
```
ENV["JULIA_PKG_SERVER"] = ""
ENV["JULIA_PKG_SERVER"] = "https://pkg.julialang.org"
```

**Comment by fredrikekre** [comment]
`ENV["JULIA_PKG_SERVER"] = ""` should work but you have to delete the registry first since otherwise you have the unsynced PkgServer registry and not a git-registry.

**Comment by fredrikekre** [comment]
Seems to be the eu server that has stopped syncing some 3 days ago. You can try setting it to e.g. `https://us-east.pkg.julialang.org` and try.

**Comment by johnnychen94** [comment]
I believe `https://us-east.storage.juliahub.com` isn't happy recently, and that most pkg servers actually uses `https://kr.storage.juliahub.com` as upstream. If your local `pkg.julialang.org` failed to fetch resources from the KR storage (e.g., HEAD timeout), then it gets outdated. I've checked that "kr.pkg.julialang.org" and "cn-east.pkg.julialang.org" contain this package version.

**Comment by visr** [comment]
`https://us-east.pkg.julialang.org` does indeed have it, `https://eu-central.pkg.julialang.org/` doesn't.

**Comment by DilumAluthge** [comment]
I'm having issues on GitHub Actions CI, which should be running on U.S. machines. E.g. from https://github.com/cscherrer/SossMLJ.jl/pull/29/checks?check_run_id=999418596 I have:
```
ERROR: Unsatisfiable requirements detected for package Soss [8ce77f84]:
 Soss [8ce77f84] log:
 ├─possible versions are: [0.1.0, 0.2.0, 0.3.0, 0.4.0, 0.5.0, 0.6.0, 0.7.0, 0.8.0, 0.9.0, 0.10.0, 0.11.0, 0.12.0] or uninstalled
 └─restricted to versions 0.13 by SossMLJ [cf2dc0df] — no versions left
   └─SossMLJ [cf2dc0df] log:
     ├─possible versions are: 0.1.0 or uninstalled
     └─SossMLJ [cf2dc0df] is fixed to version 0.1.0
```

But Soss 0.13.0 has been in the registry for two hours: https://github.com/JuliaRegistries/General/blob/master/S/Soss/Versions.toml

**Comment by DilumAluthge** [comment]
I get the same error on my local machine. I'm in Rhode Island. And typing `pkg.julialang.org` in my web browser gives me `https://us-east.pkg.julialang.org/`.

```
ERROR: Unsatisfiable requirements detected for package Soss [8ce77f84]:
 Soss [8ce77f84] log:
 ├─possible versions are: [0.1.0, 0.2.0, 0.3.0, 0.4.0, 0.5.0, 0.6.0, 0.7.0, 0.8.0, 0.9.0, 0.10.0, 0.11.0, 0.12.0] or uninstalled
 └─restricted to versions 0.13 by SossMLJ [cf2dc0df] — no versions left
   └─SossMLJ [cf2dc0df] log:
     ├─possible versions are: 0.1.0 or uninstalled
     └─SossMLJ [cf2dc0df] is fixed to version 0.1.0
```

**Comment by MilesCranmer** [comment]
My PR #29229 got merged 19 hours ago, and still it has not propagated to the registry server. I can install the Git repo version of the registry on my machine and it works, but is there any way to get this to work automatically for downstream users (in a non-Julia app that requires the latest version of my Julia package)? e.g., could Pkg automatically check the git version if it can't find the version in the registry server?

(Let me know if I should open a different issue; but seems like this is related.)

**Comment by fredrikekre** [comment]
> could Pkg automatically check the git version if it can't find the version in the registry server?

That is supposed to happen*, but something is currently not working.

**Comment by StefanKarpinski** [comment]
cc @staticfloat, can you take a look at why this is happening?

**Comment by staticfloat** [comment]
There is a flaw in the remote procedure call building blocks that we're using in the Storage Server that's causing remote workers to drop work units and eventually become desynchronized with the master.  I've just implemented a work-around that basically kicks over the workers once every 12 hours to ensure that we don't end up with complete service loss, but I've also not been able to induce these kinds of issues in any kind of stress-testing that I've tried.

**Comment by ExpandingMan** [comment]
I seem to be experiencing a manifestation of this issue.  As of tonight, I can reproduce it reliably.

I'm trying to update a package, namely AWSS3 to 0.8.3.  It's definitely there in the general registry, but is missing from the registry that Julia clones for me with 1.6-beta with pkg server enabled.  Even if I delete the registry, the entry is still missing.  I do not experience this problem if either I disable pkg server and go through git or clone the registry myself.  I have attempted to reproduce this several times over half an hour or so, and the registry cloned by pkg server was out of date every time.

I have not checked whether other entries got updated, but I would assume I'm just being given a stale registry.

**Comment by DilumAluthge** [comment]
BTW, in most use cases, this will be fixed by https://github.com/JuliaLang/Pkg.jl/pull/2377, which will be available in Julia 1.6.1.

**Comment by staticfloat** [comment]
X-ref: https://github.com/JuliaPackaging/PkgServer.jl/pull/116

**Comment by staticfloat** [comment]
This seems to have improved significantly since the work I did in February.  I'd say the only remaining thing to do would be to plot the average "time lag" and show that it no longer goes above a certain threshold, then we can close this.

**Comment by ppalmes** [comment]
yep, i just experienced this. the delay is more than an hour on this date. i'm in EU too.

**Comment by JeffFessler** [comment]
I'm seeing a delay of more than 4 hours on this registry update (MIRTjim 0.6.0):
https://github.com/JuliaRegistries/General/pull/37257
I even removed `~/.julia/registries/General` and did `] registry add General`.
Afterwords the local file `General/M/MIRTjim/Versions.toml` is behind what is on github
https://github.com/JuliaRegistries/General/blob/master/M/MIRTjim/Versions.toml

I second the suggestion that it would be nice to have some way to ping a local pkg server to find out when it was last synced.  In my case it is https://us-east2.pkg.julialang.org.
The message returned there could be more informative.

**Comment by dehann** [comment]
> It would be nice to get some concrete measurements on this.

I wasn't timing exactly but my case today on #37344 was at least 6 hours delay and counting.  Thanks for connecting me back here @DilumAluthge , I had missed this issue!

**Comment by dehann** [comment]
> I've just implemented a work-around that basically kicks over the workers once every 12 hours to ensure that we don't end up with complete service loss

One more note, the PR I'm working against (#37251) was merged more than 24 hours ago, so perhaps the case I'm seeing is one of the really slow ones.  Hopefully there isn't something more going on.  I just tried again and CI is still failing (which looks like a PkgServer sync issue).

**Comment by dehann** [comment]
Based on the fix @JeffFessler mentioned (thanks!), I was able to get this to work by just adding the environment variable `JULIA_PKG_SERVER = ""` to the CI scripts, see here:
https://github.com/JuliaRobotics/ApproxManifoldProducts.jl/blob/4a8c3acadac765191c70659e6d312d8bd38aa19c/.travis.yml#L9-L10

----

PS, without this fix I'm still getting the sync issue more than 36 hours later.

**Comment by briochemc** [comment]
Just experienced this too with a 5hr+ delay today.

**Comment by StefanKarpinski** [comment]
Yeah, it's a public resource and sometimes very large artifacts get submitted which take a long time to get processed, preventing updates for a while. If you want to see things immediately, you can set `JULIA_PKG_SERVER=""` and skip getting things from the package server altogether.

**Comment by MilesCranmer** [comment]
The delays in the registry updates have caused ~ten users of [PySR](https://github.com/MilesCranmer/PySR) (has [SymbolicRegression.jl](https://github.com/MilesCranmer/SymbolicRegression.jl) as backend) to raise GitHub issues or email me, despite me pinning an issue with the `JULIA_PKG_SERVER=""` fix and also mentioning it on my README. Can a fix be implemented if the server can't find a package's version, it automatically switches to the updated package server? Or can the git-based server be used as the new default, so users who are new to Julia don't have to debug this?

What happens is: I usually update the Julia backend, wait for it to merge with the registry, and then update the PyPI package (Python package server). PyPI is instantly updated, and my tests pass fine because the Julia GitHub action uses the git-based registry, but the Julia default registry can take more than a day sometimes, so this will cause any user (who doesn't use Julia regularly) who updates PySR to see an issue about Julia not being able to find the updated backend.

**Comment by DilumAluthge** [comment]
I madd a Discourse post that summarizes this issue and provides the workaround for users that need immediate access to new packages and new versions.

https://discourse.julialang.org/t/general-registry-delays-and-a-workaround/67537?u=dilumaluthge

**Comment by staticfloat** [comment]
With some recent upgrades to the StorageServer, this should now be pretty much fixed.  Please shout out if you experience PkgServer registry delays, as they should be eliminated now.  We have aded some client-side configuration that can be used to communicate to the PkgServer if you would like a more bleeding-edge or conservative registry, see this issue for more detail: https://github.com/JuliaPackaging/PkgServer.jl/pull/144

**Comment by waltergu** [comment]
> With some recent upgrades to the StorageServer, this should now be pretty much fixed. Please shout out if you experience PkgServer registry delays, as they should be eliminated now. We have aded some client-side configuration that can be used to communicate to the PkgServer if you would like a more bleeding-edge or conservative registry, see this issue for more detail: [JuliaPackaging/PkgServer.jl#144](https://github.com/JuliaPackaging/PkgServer.jl/pull/144)

From Julia version 1.11.1 onwards, it appears that the PkgServer registry may experience significant delays, sometimes extending to several hours.

**Comment by MilesCranmer** [comment]
Yeah it seems like there is a massive delay right now? A package registered 15 hours ago still isn't accessible to Pkg.jl.

**Comment by fredrikekre** [comment]
See https://github.com/JuliaRegistries/General/issues/120562
